### PR TITLE
Feature (SC-146): static modbus bitfield interface generation

### PR DESCRIPTION
**Pull request by gordon-epc**
## Jira Story
[SC-146](https://epcpower.atlassian.net/browse/SC-146)

## About

Add static modbus bitfield interface generation. The change requires combining the SunSpec bitfield interface generation with the static modbus bitfield interface generation. 

In addition, for this repo there were minor UI changes to handle replacement of the SunSpec specific interface generation path with the generic bitfield interface generation path.

## Associated Pull Requests

*   [ ] [embedded-library](https://github.com/epcpower/embedded-library/pull/82)
*   [ ] [grid-tied](https://github.com/epcpower/grid-tied/pull/133)

## Validation

See the grid-tied PR.

## Reproduction Steps

See the grid-tied PR.

### CODE REVIEW

File: src/epcpm/cli/sunspectostaticmodbus.py [hunk 1]
```diff
@@ -28,37 +28,57 @@ def map_type_uuid(input, sunspec_types, staticmodbus_types):
...
-        == epcpm.sunspecmodel.TableRepeatingBlockReferenceDataPointReference.__name__
-    ):
-        return create_function_data(
-            input_sunspec_data, scale_factor_uuid_map, sunspec_types, staticmodbus_types
-        )
-    else:
-        raise ValueError(
-            f"Unsupported class_name '{input_sunspec_data['class_name']}' in create_data"
-        )
+    # Create root object to store the static modbus model.
```
Comment (bholewin-epc) [inline src/epcpm/cli/sunspectostaticmodbus.py:34]: Can you create docstrings that follow the google standard? And for all the methods?

File: src/epcpm/cli/sunspectostaticmodbus.py [hunk 2]
```diff
@@ -28,37 +28,57 @@ def map_type_uuid(input, sunspec_types, staticmodbus_types):
     raise ValueError("Error: failure to map type")
 
 
-def create_data_object(
-    input_sunspec_data, scale_factor_uuid_map, sunspec_types, staticmodbus_types
+def create_data_objects(
+    input_sunspec_csv, scale_factor_uuid_map, sunspec_types, staticmodbus_types
```
Comment (bholewin-epc) [inline src/epcpm/cli/sunspectostaticmodbus.py:32]: type hints por favor.

File: src/epcpm/cli/sunspectostaticmodbus.py [hunk 3]
```diff
@@ -117,14 +138,40 @@ def create_function_data_bitfield(
...
     function_data_bitfield.size = int(input_sunspec_data["size"])
-    # A bitfield would never be a scale factor, so no need to check for uuid in scale_factor_uuid_map.
+    # A bitfield is never a scale factor, so no need to check for uuid in scale_factor_uuid_map.
     function_data_bitfield.uuid = generate_uuid()
     function_data_bitfield.address = int(input_sunspec_data["modbus_address"])
 
     return function_data_bitfield
 
 
+def create_function_data_bitfield_member(
```
Comment (bholewin-epc) [inline src/epcpm/cli/sunspectostaticmodbus.py:148]: There's a fair bit of reused code. Is it worthwhile to synthesize the data_bitfield and data_bitfield_member functions like so?:

```
def create_function_data_bitfield(input_sunspec_data, sunspec_types, staticmodbus_types, member=False):
    if member:
        function_data_bitfield = epcpm.staticmodbusmodel.FunctionDataBitfieldMember()
    else:
        function_data_bitfield = epcpm.staticmodbusmodel.FunctionDataBitfield()
    function_data_bitfield.parameter_uuid = (
        input_sunspec_data["parameter_uuid"]
        if input_sunspec_data["parameter_uuid"]
        else None
    )
    if input_sunspec_data["type_uuid"]:
        function_data_bitfield.type_uuid = map_type_uuid(
            input_sunspec_data["type_uuid"], sunspec_types, staticmodbus_types
        )
    else:
        function_data_bitfield.type_uuid = None
    if not member:
        function_data_bitfield.size = int(input_sunspec_data["size"])
        function_data_bitfield.address = int(input_sunspec_data["modbus_address"])
    else:
        function_data_bitfield.bit_offset = int(input_sunspec_data["bit_offset"])
        function_data_bitfield.bit_length = int(input_sunspec_data["bit_length"])
    function_data_bitfield.uuid = generate_uuid()
    return function_data_bitfield
```

File: src/epcpm/staticmodbustoc.py [hunk 4]
```diff
@@ -216,13 +218,18 @@ def gen(self) -> typing.List[str]:
         Returns:
             list: staticmodbusAddrRegMap rows for the generated .c file output
         """
-        # TODO: to be implemented, for now NULL for all FunctionDataBitfield objects
+        # TODO: CAMPid 9685439641536675431653179671436
```
Comment (bholewin-epc) [inline src/epcpm/staticmodbustoc.py:221]: How about a helper.py file that has some of these repeated functions?

File: src/epcpm/sunspectocsv.py [hunk 5]
```diff
@@ -419,12 +431,81 @@ def gen(self) -> typing.List[typing.List[str], int]:
...
+        # Insert the parent DataPointBitfield ahead of the list of DataPointBitfieldMember rows.
+        rows.insert(0, row)
+
+        return rows, row.size
+
+
+@builders(epcpm.sunspecmodel.DataPointBitfieldMember)
+@attr.s
+class DataPointBitfieldMember:
+    wrapped = attr.ib(type=epcpm.sunspecmodel.DataPointBitfieldMember)
```
Comment (bholewin-epc) [inline src/epcpm/sunspectocsv.py:457]: Little class docstring? :)

**Review by bholewin-epc — CHANGES_REQUESTED** [review]

File: src/epcpm/sunspectocsv.py (hunk 5 above)
Comment (gordon-epc) [reply src/epcpm/sunspectocsv.py:467]: Added class docstrings for `sunspectocsv` in f0949f07b9962d8bb4d7117c1adfd0b4ccae0de9

File: src/epcpm/cli/sunspectostaticmodbus.py (hunk 1 above)
Comment (gordon-epc) [reply src/epcpm/cli/sunspectostaticmodbus.py]: Added docstrings in 73afd44b3cf2d463467b95ab4664543c5ee732e7

File: src/epcpm/cli/sunspectostaticmodbus.py (hunk 2 above)
Comment (gordon-epc) [reply src/epcpm/cli/sunspectostaticmodbus.py]: Added type hints in 73afd44b3cf2d463467b95ab4664543c5ee732e7

File: src/epcpm/cli/sunspectostaticmodbus.py (hunk 3 above)
Comment (gordon-epc) [reply src/epcpm/cli/sunspectostaticmodbus.py:226]: Refactored the three `create_function_data_*` methods in 73afd44b3cf2d463467b95ab4664543c5ee732e7

File: src/epcpm/staticmodbustoc.py (hunk 4 above)
Comment (gordon-epc) [reply src/epcpm/staticmodbustoc.py]: I added `pm_helper.py` and refactored `CAMPid 9685439641536675431653179671436` and `CAMPid 07954360685417610543064316843160`. (Note: both of these CAMPid's turned out to be the same.) In addition, there were spots that were not marked with either CAMPid that also needed to be refactored. Fixed in 6091c44c4e0c921e0eaa944b9ddfff75e11dcdd4

I'm not going to do any more of these CAMPid refactors in this PR as each one is a rabbit hole. I recommend they are addressed in a future PR (or PR's).

Comment (bholewin-epc) [reply src/epcpm/staticmodbustoc.py]: Totally agree

**Review by bholewin-epc — APPROVED** [review]

File: src/epcpm/sunspectocsv.py [hunk 6]
```diff
@@ -419,17 +439,90 @@ def gen(self) -> typing.List[typing.List[str], int]:
...
+                parameter_uuid_finder=self.parameter_uuid_finder,
+                model_id=self.model_id,
+                model_offset=self.model_offset,
+                address_offset=self.address_offset,
+            )
+            child_row = builder.gen()
+            rows.append(child_row)
+
+        # Insert the parent DataPointBitfield ahead of the list of DataPointBitfieldMember rows.
+        rows.insert(0, row)
```
Comment (epcAnker) [inline src/epcpm/sunspectocsv.py:457]: replace 0 with something like START_OF_LIST_INDEX ?

Comment (bholewin-epc) [reply src/epcpm/sunspectocsv.py:457]: Just my 2 cents but I am cool with 0's since that's generally a given for start of array/list index.

Comment (gordon-epc) [reply src/epcpm/sunspectocsv.py]: While the insert(0, x) is a common way to insert at the front of a list, there is a little smell here. I didn't have to wait until the end to add the first row. I moved the add prior to the rest of the rows being added. See fc8890ed79698211e5fc27103b241f25a4ba60c9

File: src/epcpm/importexportdialog.ui [hunk 7]
```diff
@@ -101,40 +101,40 @@
...
       <string>...</string>
      </property>
     </widget>
    </item>
    <item row="5" column="0">
-    <widget class="QLabel" name="sunspec_bitfields_c_label">
+    <widget class="QLabel" name="sunspec_tables_c_label">
      <property name="text">
-      <string>SunSpec Bitfields C</string>
+      <string>SunSpec Tables C:</string>
```
Comment (epcAnker) [inline src/epcpm/importexportdialog.ui:126]: should the earlier version have had a ":" also?  looks like that is consistent with the other code.

File: src/epcpm/parameterstobitfieldsc.py [hunk 8]
```diff
@@ -0,0 +1,504 @@
...
+
+                c_lines.extend(members_interface_c_lines)
+
+                # Generate the combined interface item code
+                c_lines.extend(
+                    [
+                        f"InterfaceItem_Bitfield interfaceItem_{name_uuid} = {{",
+                        [
+                            f".common = {{",
+                            [ccl for ccl in common_c_lines],
```
Comment (epcAnker) [inline src/epcpm/parameterstobitfieldsc.py:211]: rename ccl to common\_c\_line ?

File: src/epcpm/parameterstobitfieldsc.py [hunk 9]
```diff
@@ -0,0 +1,504 @@
...
+                            f"}},",
+                        ],
+                        [micl for micl in members_info_c_lines],
```
Comment (epcAnker) [inline src/epcpm/parameterstobitfieldsc.py:214]: rename micl to members\_info\_c\_line?

File: src/epcpm/parameterstobitfieldsc.py [hunk 10]
```diff
@@ -0,0 +1,504 @@
...
+
+    def gen_common_interface(self) -> typing.List[str]:
+        """
+        Generate a SunSpec bitfield common definitions for the interface item.
+
+        Returns:
+            list: bitfield common definitions
+        """
+        bits_per_modbus_register = 16
+        bit_length = bits_per_modbus_register * self.wrapped.size
```
Comment (epcAnker) [inline src/epcpm/parameterstobitfieldsc.py:423]: should this be renamed to bitfield\_bit\_length?

**Review by epcAnker — APPROVED** [review]
look at comments

File: src/epcpm/importexportdialog.ui (hunk 7 above)
Comment (gordon-epc) [reply src/epcpm/importexportdialog.ui:126]: Yes. All the other labels end with a colon, but for some reason this one didn't have it, so I added it in to be consistent.

File: src/epcpm/parameterstobitfieldsc.py (hunk 8 above)
Comment (gordon-epc) [reply src/epcpm/parameterstobitfieldsc.py:211]: I was considering going the other way and changing the code like this:
`[c for c in common_c_lines],`

If I change `ccl` to `common_c_line`, black reformats this one liner to four lines. It's already a somewhat difficult to read code section and increasing the variable name length makes it worse, IMO.

Also note that this one liner is a common python syntax called "list comprehension".

File: src/epcpm/parameterstobitfieldsc.py (hunk 9 above)
Comment (gordon-epc) [reply src/epcpm/parameterstobitfieldsc.py:214]: See [this comment](https://github.com/epcpower/pm/pull/25#discussion_r749690952).

File: src/epcpm/parameterstobitfieldsc.py (hunk 10 above)
Comment (gordon-epc) [reply src/epcpm/parameterstobitfieldsc.py:423]: No, I used the variable name `bit_length` to match other code that uses the same name. Consistency.

Comment (gordon-epc) [reply src/epcpm/parameterstobitfieldsc.py:420]: Changed in 7c020f4c2743bae15875da13eb0486b3427ac3a5

Also removed some unused code that was somewhat related.

File: src/epcpm/parameterstobitfieldsc.py (hunk 8 above)
Comment (epcAnker) [reply src/epcpm/parameterstobitfieldsc.py:211]: sounds good.  yea list comprehension is confusing

File: src/epcpm/parameterstobitfieldsc.py (hunk 10 above)
Comment (epcAnker) [reply src/epcpm/parameterstobitfieldsc.py:420]: i think there are other mentions of bit\_length.  Should they all be change bitfield\_bit\_length?

Comment (gordon-epc) [reply src/epcpm/parameterstobitfieldsc.py:420]: No, the other `bit_length` variables have to do with the length of bitfield members and those are consistently using `bit_length` in several places. The ones I just changed have to do with the bitfield length (16 or 32 in this case).

**Pull request merged**